http-server/src/main/java/org/triplea/server/moderator/toolbox/api/key/validation/ApiKeyValidationService.java

```diff
@@ -54,34 +58,46 @@ public void verifyApiKey(final HttpServletRequest request) {
    *
    * @throws IllegalArgumentException Thrown if the http servlet request headers do not contain a moderator api key.
    * @throws IncorrectApiKeyException Thrown if the provided API key does not match any known keys.
-   * @throws ApiKeyVerificationLockOutException Thrown if rate-limiting has kicked in and the API key verification
+   * @throws ApiKeyLockOutException Thrown if rate-limiting has kicked in and the API key verification
    *         was not attempted. This can be caused by too many attempts from a specific IP address, or too many failed
    *         attempts across all IP addresses. To avoid the latter case from locking all users out, valid API keys
    *         are cached and will by-pass the lock-out if we enter into that state. We do a lock-out for all
    *         IP addresses in case an attacker sets up many computers or is able to spoof their IP address.
    */
   public synchronized int lookupModeratorIdByApiKey(final HttpServletRequest request) {
-    final String apiKey = request.getHeader(ModeratorToolboxClient.MODERATOR_API_KEY_HEADER);
-    Preconditions.checkArgument(apiKey != null && !apiKey.isEmpty());
+    final String hashedKey = extractHashedKey(request);
```

Comment:
This is another important location, it was where we previously did api key verification. Now we need to combine password for a salt. Some logging was also added.

Reply:
I'm wondering: Is there any chance this code might be vulnerable to a timing attack?
In other words: Is user data always being hashed before being compared in memory or against the Database or something?
I haven't really looked into the algorithm of yours in-depth, so both might be possible.

Reply:
I don't think so, notably because there is not a second piece of information being validated, just the key.

The primary defense is the lockout feature. Ignoring that, we always hash the key with password, or in the single-use case hash the single-use key and compare to what is in DB. The lookup time of bad key vs good key will be a function of database retrieval time. If a user has a good key, then they have a good key, bad key attempts will always take the same amount of time (assuming DB lookup time remains pretty speedy no matter what kind of key value is used).